Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -149,4 +165,405 @@ theorem rootsOfUnity_eq_torsion [NumberField K] :
+local macro_rules | `($x ^ $y) => `(HPow.hPow $x $y)
+
+variable (w₁ : InfinitePlace K) {B : ℕ} (hB : minkowskiBound K < (convexBodyLtFactor K) * B)
+
+/-- This result shows that there always exists a next term in the sequence. -/
+theorem seq_next {x : 𝓞 K} (hx : x ≠ 0) :
+    ∃ y : 𝓞 K, y ≠ 0 ∧ (∀ w, w ≠ w₁ → w y < w x) ∧ |Algebra.norm ℚ (y : K)| ≤ B := by
+  let f : InfinitePlace K → ℝ≥0 :=
+    fun w => ⟨(w x) / 2, div_nonneg (AbsoluteValue.nonneg _ _) (by norm_num)⟩
+  suffices ∀ w, w ≠ w₁ → f w ≠ 0 by
+    obtain ⟨g, h_geqf, h_gprod⟩ := adjust_f K B this
+    obtain ⟨y, h_ynz, h_yle⟩ := exists_ne_zero_mem_ringOfIntegers_lt (f := g)
+      (by rw [convexBodyLt_volume]; convert hB; exact congr_arg ((↑): NNReal → ENNReal) h_gprod)
+    refine ⟨y, h_ynz, fun w hw => (h_geqf w hw ▸ h_yle w).trans ?_, ?_⟩
+    · rw [← Rat.cast_le (K := ℝ), Rat.cast_coe_nat]
+      calc
+        _ = ∏ w : InfinitePlace K, w y ^ mult w := (prod_eq_abs_norm (y : K)).symm
+        _ ≤ ∏ w : InfinitePlace K, (g w : ℝ) ^ mult w := by
+          refine Finset.prod_le_prod ?_ ?_
+          · exact fun _ _ => pow_nonneg (by positivity) _
+          · exact fun w _ => pow_le_pow_of_le_left (by positivity) (le_of_lt (h_yle w)) (mult w)
+        _ ≤ (B : ℝ) := by
+          simp_rw [← NNReal.coe_pow, ← NNReal.coe_prod]
+          exact le_of_eq (congr_arg toReal h_gprod)
+    · refine div_lt_self ?_ (by norm_num)
+      simp only [pos_iff, ne_eq, ZeroMemClass.coe_eq_zero, hx]
+  intro _ _
+  rw [ne_eq, Nonneg.mk_eq_zero, div_eq_zero_iff, map_eq_zero, not_or, ZeroMemClass.coe_eq_zero]
+  exact ⟨hx, by norm_num⟩
+
+/-- An infinite sequence of nonzero algebraic integers of `K` satisfying the following properties:
+• `seq n` is nonzero;
+• for `w : InfinitePlace K`, `w ≠ w₁ → w (seq n+1) < w (seq n)`;
+• `∣norm (seq n)∣ ≤ B`. -/
+def seq : ℕ → { x : 𝓞 K // x ≠ 0 }
+  | 0 => ⟨1, by norm_num⟩
+  | n + 1 =>
+    ⟨(seq_next K w₁ hB (seq n).prop).choose, (seq_next K w₁ hB (seq n).prop).choose_spec.1⟩
+
+/-- The terms of the sequence are nonzero. -/
+theorem seq_ne_zero (n : ℕ) : (seq K w₁ hB n : K) ≠ 0 := by
+  refine (map_ne_zero_iff (algebraMap (𝓞 K) K) ?_).mpr (seq K w₁ hB n).prop
+  exact IsFractionRing.injective { x // x ∈ 𝓞 K } K
+
+/-- The sequence is strictly decreasing at infinite places distinct from `w₁`. -/
+theorem seq_antitone {n m : ℕ} (h : n < m) :
+    ∀ w : InfinitePlace K, w ≠ w₁ → w (seq K w₁ hB m) < w (seq K w₁ hB n) := by
+  induction m with
+  | zero =>
+      exfalso
+      exact Nat.not_succ_le_zero n h
+  | succ m m_ih =>
+      intro w hw
+      cases eq_or_lt_of_le (Nat.le_of_lt_succ h) with
+      | inl hr =>
+          rw [hr]
+          exact (seq_next K w₁ hB (seq K w₁ hB m).prop).choose_spec.2.1 w hw
+      | inr hr =>
+          refine lt_trans ?_ (m_ih hr w hw)
+          exact (seq_next K w₁ hB (seq K w₁ hB m).prop).choose_spec.2.1 w hw
+
+/-- The terms of the sequence have norm bounded by `B`. -/
+theorem seq_norm_bdd (n : ℕ) :
```

Comment:
We usually state two separate results rather than using `∧`.

Reply:
I split it into two lemmas.